File: dynamic-map/src/utils/reactiveAlgorithm.ts
import { generateId } from './id'
import {
  buildLevelSequence,
  kindForLevel,
  labelFor,
  layoutSchedule,
  pitstopCountFor,
  randomCoordinate,
} from './pacingAlgorithm'
import type { CognitiveLevel, PacingPath, Pitstop, Season, Task } from '../types/cognitiveHub'

const MIN_REGENERATION_MS = 90_000

const GROUNDING_PITSTOPS_BY_SEASON: Partial<Record<Season, number>> = {
  'Exhaustion v2': 2,
  Grounding: 2,
  Doubt: 1,
  'Frustration Season': 1,
}

function groundingCountFor(season: Season): number {
  return GROUNDING_PITSTOPS_BY_SEASON[season] ?? 0
}

/**
 * Handles a "stranded" deviation: locks in everything scheduled before `currentTime`,
 * drops a stranding marker at the point of deviation, and regenerates only the future
 * portion of the path so it fits the new season while preserving the original total
 * session duration.
 */
export function handleDeviation(
  currentTime: number,
  newSeason: Season,
  existingPath: PacingPath,
  task: Task,
): PacingPath {
  const totalDurationMs = existingPath.totalDurationMs
  const pastPath = existingPath.pitstops.filter((p) => p.scheduledTime < currentTime)
  const placedCoordinates = pastPath.map((p) => p.mapCoordinates)

  const strandingMarker: Pitstop = {
    id: generateId('pitstop'),
    label: labelFor('stranding', 1),
    level: 1,
    season: newSeason,
    kind: 'stranding',
    scheduledTime: currentTime,
    realTimeTrigger: Date.now(),
    mapCoordinates: randomCoordinate(placedCoordinates),
  }
  placedCoordinates.push(strandingMarker.mapCoordinates)

  const remainingMs = Math.max(totalDurationMs - currentTime, 0)
  const futurePitstops: Pitstop[] = []

  if (remainingMs <= MIN_REGENERATION_MS) {
    if (remainingMs > 0) {
      const mapCoordinates = randomCoordinate(placedCoordinates)
      futurePitstops.push({
        id: generateId('pitstop'),
        label: labelFor('end', 2),
        level: 2,
        season: newSeason,
        kind: 'end',
        scheduledTime: totalDurationMs,
        mapCoordinates,
      })
    }
  } else {
    const remainingMinutes = remainingMs / 60_000
    const futureCount = Math.max(2, pitstopCountFor(remainingMinutes, task.rateOfChange))
    const groundingCount = Math.min(groundingCountFor(newSeason), futureCount - 1)

    // +1 so we can drop the leading point (which would land exactly on currentTime).
    const schedule = layoutSchedule(futureCount + 1, currentTime, remainingMs).slice(1)
    const oscillationLevels = buildLevelSequence(
      futureCount,
      task.cognitiveLimit,
      newSeason,
      task.rateOfChange,
    )

    for (let i = 0; i < futureCount; i++) {
      const isLast = i === futureCount - 1
      const isGrounding = i < groundingCount
      const level: CognitiveLevel = isLast
        ? oscillationLevels[i]
        : isGrounding
          ? ((i % 2 === 0 ? 1 : 2) as CognitiveLevel)
          : oscillationLevels[i]
      const kind = isLast
        ? 'end'
        : isGrounding
          ? 'grounding'
          : kindForLevel(1, 3, level, task.cognitiveLimit, newSeason) // mid-range index/count => never 'start'/'end'
      const mapCoordinates = randomCoordinate(placedCoordinates)
      placedCoordinates.push(mapCoordinates)

      futurePitstops.push({
        id: generateId('pitstop'),
        label: labelFor(kind, level),
        level,
        season: newSeason,
        kind,
        scheduledTime: schedule[i],
        mapCoordinates,
      })
    }
  }

  return {
    pitstops: [...pastPath, strandingMarker, ...futurePitstops],
    totalDurationMs,
  }
}

File: watch-app/CognitivePacingAtlasWatch/Models/ReactiveAlgorithm.swift
import Foundation

// Swift port of dynamic-map/src/utils/reactiveAlgorithm.ts.
enum ReactiveAlgorithm {
    private static let minRegenerationMs = 90_000.0

    private static let groundingPitstopsBySeason: [Season: Int] = [
        .exhaustionV2: 2,
        .grounding: 2,
        .doubt: 1,
        .frustrationSeason: 1,
    ]

    private static func groundingCount(for season: Season) -> Int {
        groundingPitstopsBySeason[season] ?? 0
    }

    /// Handles a "stranded" deviation: locks in everything scheduled before `currentTime`,
    /// drops a stranding marker at the point of deviation, and regenerates only the future
    /// portion of the path so it fits the new season while preserving the original total
    /// session duration.
    static func handleDeviation(
        currentTime: Double,
        newSeason: Season,
        existingPath: PacingPath,
        task: WatchTask,
        preferences: UserPreferences
    ) -> PacingPath {
        let totalDurationMs = existingPath.totalDurationMs
        let pastPath = existingPath.pitstops.filter { $0.scheduledTime < currentTime }
        var placedCoordinates = pastPath.map(\.mapCoordinates)
        let minLevel = min(task.minCognitiveLevel, task.cognitiveLimit)
        let maxLevel = task.cognitiveLimit

        let strandingMarker = Pitstop(
            id: generateId("pitstop"),
            label: PacingAlgorithm.labelFor(.stranding, 1),
            level: 1,
            season: newSeason,
            kind: .stranding,
            scheduledTime: currentTime,
            realTimeTrigger: Date().timeIntervalSince1970 * 1000,
            mapCoordinates: PacingAlgorithm.randomCoordinate(existing: placedCoordinates)
        )
        placedCoordinates.append(strandingMarker.mapCoordinates)

        let remainingMs = max(totalDurationMs - currentTime, 0)
        var futurePitstops: [Pitstop] = []

        if remainingMs <= minRegenerationMs {
            if remainingMs > 0 {
                let coords = PacingAlgorithm.randomCoordinate(existing: placedCoordinates)
                futurePitstops.append(Pitstop(
                    id: generateId("pitstop"),
                    label: PacingAlgorithm.labelFor(.end, minLevel),
                    level: minLevel,
                    season: newSeason,
                    kind: .end,
                    scheduledTime: totalDurationMs,
                    realTimeTrigger: nil,
                    mapCoordinates: coords
                ))
            }
        } else {
            let remainingMinutes = remainingMs / 60_000
            let futureCount = max(
                2, PacingAlgorithm.pitstopCount(forDurationMinutes: remainingMinutes, rateOfChange: task.rateOfChange)
            )
            let groundCount = min(groundingCount(for: newSeason), futureCount - 1)

            // +1 so we can drop the leading point (which would land exactly on currentTime).
            let fullSchedule = PacingAlgorithm.layoutSchedule(
                count: futureCount + 1, startMs: currentTime, durationMs: remainingMs,
                splitType: preferences.splitType
            )
            let schedule = Array(fullSchedule.dropFirst())
            let oscillationLevels = PacingAlgorithm.buildLevelSequence(
                count: futureCount, minLevel: minLevel, maxLevel: maxLevel,
                entrySeason: newSeason, rateOfChange: task.rateOfChange,
                cyclePattern: preferences.cyclePattern
            )

            for i in 0..<futureCount {
                let isLast = i == futureCount - 1
                let isGrounding = i < groundCount
                let level: CognitiveLevel = isLast
                    ? oscillationLevels[i]
                    : isGrounding
                        ? (i % 2 == 0 ? minLevel : min(minLevel + 1, maxLevel))
                        : oscillationLevels[i]
                // mid-range index/count (1 of 3) so kindForLevel never returns 'start'/'end'.
                let kind: PitstopKind = isLast
                    ? .end
                    : isGrounding
                        ? .grounding
                        : PacingAlgorithm.kindForLevel(
                            index: 1, count: 3, level: level,
                            minLevel: minLevel, maxLevel: maxLevel, entrySeason: newSeason
                        )
                let coords = PacingAlgorithm.randomCoordinate(existing: placedCoordinates)
                placedCoordinates.append(coords)

                futurePitstops.append(Pitstop(
                    id: generateId("pitstop"),
                    label: PacingAlgorithm.labelFor(kind, level),
                    level: level,
                    season: newSeason,
                    kind: kind,
                    scheduledTime: schedule[i],
                    realTimeTrigger: nil,
                    mapCoordinates: coords
                ))
            }
        }

        return PacingPath(
            pitstops: pastPath + [strandingMarker] + futurePitstops,
            totalDurationMs: totalDurationMs
        )
    }
}
